eZ/Bundle/EzPublishCoreBundle/FieldType/RichText/Converter/Html5Input.php

```diff
@@ -22,4 +23,27 @@ public function __construct($stylesheet, ConfigResolverInterface $configResolver
         $customStylesheets = $customStylesheets ?: array();
         parent::__construct($stylesheet, $customStylesheets);
     }
+
+    /**
+     * Method checks if document contains any links with '[' or ']' sign an replaces it to its html entities version.
+     *
+     * @param DOMDocument $document
+     *
+     * @return DOMDocument
+     */
+    public function convert(DOMDocument $document)
+    {
+        // EZP-27562 - links containing [ or ] sign results in xml validation error
+        foreach ($document->getElementsByTagName('a') as $link) {
+            $href = $link->getAttribute('href');
+            if ($href !== '') {
+                $href = str_replace('[', '%5B', $href);
+                $href = str_replace(']', '%5D', $href);
```

Comment:
Hi @alongosz 

> // btw. I still have a weird feeling we're reinventing the wheel, but haven't found anything better to move this forward.

I think your hunch is good, because this URL is not being stored, what's really stored is internal reference to the stored URL, like `ezurl://42`. The problem is this happens is [external storage](https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/FieldType/RichText/RichTextStorage.php), which is **after** validation.

I think we have some design issues here, we also miss `UrlService` I hinted about on few places before:

- https://github.com/ezsystems/ezpublish-kernel/blob/e2d26faa54715675fc8f1eef3defbe192973bf5b/eZ/Publish/Core/FieldType/RichText/Converter/Render/Embed.php#L194
- https://github.com/ezsystems/ezpublish-kernel/blob/e2d26faa54715675fc8f1eef3defbe192973bf5b/eZ/Publish/Core/FieldType/RichText/Converter/Link.php#L121